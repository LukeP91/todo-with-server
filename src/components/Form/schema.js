import * as yup from 'yup'

const schema = yup.object().shape({
  title: yup
    .string()
    .min(6, 'Title has to be longer than 6 characters!')
    .max(20, 'Title has to be shorter than 20 characters!')
    .required('Title is required!'),
  description: yup
    .string()
    .min(15, 'Description has to be longer than 6 characters!')
    .max(300, 'Description has to be shorter than 20 characters!')
    .required('Description is required!'),
})

export default schema
