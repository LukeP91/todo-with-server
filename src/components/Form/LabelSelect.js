import _ from 'lodash'
import React from 'react'
import PropTypes from 'prop-types'
import { withStyles } from '@material-ui/core/styles'
import InputLabel from '@material-ui/core/InputLabel'
import MenuItem from '@material-ui/core/MenuItem'
import FormControl from '@material-ui/core/FormControl'
import Select from '@material-ui/core/Select'
import Chip from '@material-ui/core/Chip'

const styles = theme => ({
  root: {
    display: 'flex',
    flexWrap: 'wrap',
  },
  formControl: {
    margin: theme.spacing.unit,
    minWidth: 120,
    maxWidth: 300,
  },
  chips: {
    display: 'flex',
    flexWrap: 'wrap',
  },
  chip: {
    margin: theme.spacing.unit / 4,
  },
})

const ITEM_HEIGHT = 48
const ITEM_PADDING_TOP = 8
const MenuProps = {
  PaperProps: {
    style: {
      maxHeight: ITEM_HEIGHT * 4.5 + ITEM_PADDING_TOP,
      width: 250,
    },
  },
}

class MultipleSelect extends React.Component {
  selected = () => {
    const {
      labels,
      classes,
      form: { values },
    } = this.props
    return labels
      .filter(label => _.includes(values.labelsIds, label.id))
      .map(label => <Chip className={classes.chip} key={label.id} label={label.value} />)
  }

  render() {
    const {
      field: { name, ...field },
      theme,
      classes,
      labels,
      form: { values },
    } = this.props

    return (
      <div className={classes.root}>
        <FormControl className={classes.formControl} id="labels-list">
          <InputLabel htmlFor="select-multiple-chip">Labels</InputLabel>
          <Select
            {...field}
            id={name}
            MenuProps={MenuProps}
            multiple
            name={name}
            renderValue={() => <div className={classes.chips}>{this.selected()}</div>}
          >
            {labels.map(label => (
              <MenuItem
                key={label.id}
                style={{
                  fontWeight:
                    values.labelsIds.indexOf(label.id) === -1
                      ? theme.typography.fontWeightRegular
                      : theme.typography.fontWeightMedium,
                }}
                value={label.id}
              >
                {label.value}
              </MenuItem>
            ))}
          </Select>
        </FormControl>
      </div>
    )
  }
}

MultipleSelect.propTypes = {
  classes: PropTypes.object.isRequired,
  field: PropTypes.object,
  labels: PropTypes.array.isRequired,
  theme: PropTypes.object,
}

export default withStyles(styles, { withTheme: true })(MultipleSelect)
