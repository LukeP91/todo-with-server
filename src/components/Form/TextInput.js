import _ from 'lodash'
import { withStyles } from '@material-ui/core/styles'
import FormControl from '@material-ui/core/FormControl'
import FormHelperText from '@material-ui/core/FormHelperText'
import Input from '@material-ui/core/Input'
import InputLabel from '@material-ui/core/InputLabel'
import PropTypes from 'prop-types'
import React, { Component } from 'react'

const styles = {
  container: {
    marginBottom: '20px',
  },
}

class TextInput extends Component {
  renderInput = props => {
    const {
      field: { name, ...field },
      form: { errors },
      label,
      classes,
    } = props

    if (_.isEmpty(errors)) {
      return (
        <FormControl className={classes.container}>
          <InputLabel htmlFor="name-simple">{label}</InputLabel>
          <Input id="name-simple" {...field} name={name} />
        </FormControl>
      )
    } else {
      return (
        <FormControl aria-describedby="name-error-text" className={classes.container} error>
          <InputLabel htmlFor="name-error">{label}</InputLabel>
          <Input id="name-error" {...field} name={name} />
          <FormHelperText id="name-error-text">{errors[name]}</FormHelperText>
        </FormControl>
      )
    }
  }

  render() {
    return this.renderInput(this.props)
  }
}

TextInput.propTypes = {
  field: PropTypes.object,
  form: PropTypes.object,
  label: PropTypes.string,
}

export default withStyles(styles)(TextInput)
