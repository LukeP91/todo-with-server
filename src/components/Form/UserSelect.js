import { withStyles } from '@material-ui/core/styles'
import FormControl from '@material-ui/core/FormControl'
import InputLabel from '@material-ui/core/InputLabel'
import MenuItem from '@material-ui/core/MenuItem'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import Select from '@material-ui/core/Select'

const styles = {
  container: {
    marginBottom: '20px',
  },
}

class TextInput extends Component {
  renderInput = props => {
    const {
      field: { name, ...field },
      label,
      classes,
      users,
    } = props

    return (
      <FormControl className={classes.container} id="users-list">
        <InputLabel htmlFor="name-simple">{label}</InputLabel>
        <Select {...field} id="user-select" name={name}>
          {users.map(user => (
            <MenuItem key={user.id} value={user.id}>
              {user.name}
            </MenuItem>
          ))}
        </Select>
      </FormControl>
    )
  }

  render() {
    return this.renderInput(this.props)
  }
}
TextInput.propTypes = {
  classes: PropTypes.object,
  field: PropTypes.object,
  form: PropTypes.object,
  label: PropTypes.string,
  users: PropTypes.array,
}

export default withStyles(styles)(TextInput)
