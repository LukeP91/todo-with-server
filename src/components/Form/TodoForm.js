import { Field, Form } from 'formik'
import { withStyles } from '@material-ui/core/styles'
import Button from '@material-ui/core/Button'
import PropTypes from 'prop-types'
import React from 'react'

import LabelSelect from './LabelSelect'
import TextInput from './TextInput'
import UserSelect from './UserSelect'

const styles = {
  container: {
    display: 'flex',
    flexWrap: 'wrap',
    flexDirection: 'column',
  },
}

const TodoForm = ({ classes, labels, users }) => (
  <Form className={classes.container}>
    <Field component={TextInput} label="Title" name="title" />
    <Field component={TextInput} label="Description" name="description" />
    <Field component={UserSelect} label="User" name="userId" users={users} />
    <Field component={LabelSelect} label="Labels" labels={labels} name="labelsIds" />
    <Button color="primary" id="submit-button" type="submit" variant="contained">
      Submit
    </Button>
  </Form>
)

TodoForm.propTypes = {
  classes: PropTypes.object.isRequired,
  labels: PropTypes.array,
  users: PropTypes.array,
}

export default withStyles(styles)(TodoForm)
