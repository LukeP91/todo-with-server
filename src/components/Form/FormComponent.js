import { Formik } from 'formik'
import PropTypes from 'prop-types'
import React, { Component } from 'react'

import schema from './schema'
import TodoForm from './TodoForm'

export default class FormComponent extends Component {
  render() {
    const { users, labels, initialValues, submit } = this.props

    return (
      <Formik
        initialValues={initialValues}
        onSubmit={submit}
        render={formikProps => <TodoForm {...formikProps} labels={labels} users={users} />}
        validationSchema={schema}
      />
    )
  }
}

FormComponent.propTypes = {
  initialValues: PropTypes.object,
  labels: PropTypes.array,
  submit: PropTypes.func,
  users: PropTypes.array,
}

