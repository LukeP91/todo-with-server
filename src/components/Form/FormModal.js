import { connect } from 'react-redux'
import { withStyles } from '@material-ui/core/styles'
import Button from '@material-ui/core/Button'
import Modal from '@material-ui/core/Modal'
import PropTypes from 'prop-types'
import React from 'react'

import { addTodo } from '../../store/reducers/todos/actions'
import { fetchLabels } from '../../store/reducers/labels/actions'
import { fetchUsers } from '../../store/reducers/users/actions'
import Form from './FormComponent'

const styles = theme => ({
  button: {},
  container: {
    display: 'inline',
  },
  fab: {
    bottom: theme.spacing.unit * 4,
    position: 'absolute',
    right: theme.spacing.unit * 4,
  },
  modal: {
    backgroundColor: theme.palette.background.paper,
    boxShadow: theme.shadows[5],
    left: '50%',
    padding: theme.spacing.unit * 4,
    position: 'absolute',
    top: '50%',
    transform: 'translate(-50%, -50%)',
    width: theme.spacing.unit * 50,
  },
})

export class SimpleModal extends React.Component {
  state = {
    open: false,
  }

  componentDidMount() {
    this.props.fetchUsers('http://localhost:3001/users', { method: 'GET' })
    this.props.fetchLabels('http://localhost:3001/labels', { method: 'GET' })
  }

  handleOpen = () => {
    this.setState({ open: true })
  }

  handleClose = () => {
    this.setState({ open: false })
  }

  submit = (values, actions) => {
    const { submit } = this.props
    submit(values, actions)
    this.handleClose()
  }

  render() {
    const {
      classes,
      color,
      buttonLabel,
      variant,
      fab,
      initialValues,
      users,
      labels,
      id,
    } = this.props

    return (
      <div className={classes.container}>
        <Button
          className={fab ? classes.fab : classes.button}
          color={color}
          id={id}
          onClick={this.handleOpen}
          variant={variant}
        >
          {buttonLabel}
        </Button>
        <Modal onClose={this.handleClose} open={this.state.open}>
          <div className={classes.modal}>
            <Form
              handleClose={this.handleClose}
              initialValues={initialValues}
              labels={labels}
              submit={this.submit}
              users={users}
            />
          </div>
        </Modal>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  labels: state.labels.labels,
  users: state.users.users,
})

const mapDispatchToProps = {
  addTodo,
  fetchLabels,
  fetchUsers,
}

SimpleModal.propTypes = {
  buttonLabel: PropTypes.any,
  classes: PropTypes.object.isRequired,
  color: PropTypes.string,
  fab: PropTypes.bool,
  fetchLabels: PropTypes.func,
  fetchUsers: PropTypes.func,
  id: PropTypes.number,
  initialValues: PropTypes.object.isRequired,
  labels: PropTypes.array,
  submit: PropTypes.func,
  users: PropTypes.array,
  variant: PropTypes.string,
}

export default withStyles(styles)(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  )(SimpleModal),
)
