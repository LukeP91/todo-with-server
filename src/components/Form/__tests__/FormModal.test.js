import React from 'react'
import renderer from 'react-test-renderer'

import { SimpleModal } from '../FormModal'
import Form from '../FormComponent'

describe('SimpleModal', () => {
  const defaultProps = {
    buttonLabel: 'test',
    classes: {},
    color: 'primary',
    fetchLabels: jest.fn(),
    fetchUsers: jest.fn(),
    initialValues: { title: '', description: '', userId: '', labelsIds: [] },
    labels: [{ id: 1, value: 'test' }],
    users: [{ id: 1, name: 'Joe Doe' }],
    submit: jest.fn(),
  }

  const setup = overrideProps => {
    const props = { ...defaultProps, ...overrideProps }
    const wrapper = renderer.create(<SimpleModal {...props} />)
    return { props, wrapper }
  }

  it('renders correctly', () => {
    const { wrapper, props } = setup()

    expect(wrapper).toMatchSnapshot()
    expect(props.fetchLabels).toBeCalledWith('http://localhost:3001/labels', { method: 'GET' })
    expect(props.fetchUsers).toBeCalledWith('http://localhost:3001/users', { method: 'GET' })
  })

  xit('opens renders modal with form', () => {
    const { wrapper } = setup()
    expect(wrapper.getInstance().state.open).toEqual(false)
    const button = wrapper.root.findByType('button')
    button.props.onClick()

    expect(wrapper.getInstance().state.open).toEqual(true)
  })

  xit('on submit it closes the modal', () => {
    const { wrapper, props } = setup()
    wrapper.find().prop('onClick')()
    const form = wrapper.find(Form)
    form.prop('submit')()

    expect(props.submit).toBeCalled()
    expect(wrapper.state().open).toEqual(false)
  })
})
