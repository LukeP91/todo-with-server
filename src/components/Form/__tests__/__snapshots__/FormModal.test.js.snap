// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleModal renders correctly 1`] = `
<div>
  <button
    className="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-text-3 MuiButton-textPrimary-4 MuiButton-flat-6 MuiButton-flatPrimary-7"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-2"
    >
      test
    </span>
    <span
      className="MuiTouchRipple-root-30"
    />
  </button>
</div>
`;
